File: src/main/java/me/lokka30/levelledmobs/rules/strategies/LevellingStrategy.java
/*
 * Copyright (c) 2020-2021  lokka30. Use of this source code is governed by the GNU AGPL v3.0 license that can be found in the LICENSE.md file.
 */

package me.lokka30.levelledmobs.rules.strategies;

import me.lokka30.levelledmobs.misc.LivingEntityWrapper;

/**
 * Interface for the various levelling systems
 *
 * @author lokka30
 * @since 3.0.0
 */
public interface LevellingStrategy {

    int generateLevel(final LivingEntityWrapper lmEntity, final int minLevel, final int maxLevel);

    void mergeRule(final LevellingStrategy levellingStrategy);

    LevellingStrategy cloneItem();
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/rules/strategies/LevellingStrategy.kt
package io.github.arcaneplugins.levelledmobs.rules.strategies

import io.github.arcaneplugins.levelledmobs.wrappers.LivingEntityWrapper

/**
 * Interface for the various levelling systems
 *
 * @author lokka30
 * @since 3.0.0
 */
interface LevellingStrategy {
    fun generateNumber(
        lmEntity: LivingEntityWrapper,
        minLevel: Int,
        maxLevel: Int
    ): Float

    fun mergeRule(levellingStrategy: LevellingStrategy?)

    fun cloneItem(): LevellingStrategy

    val strategyType: StrategyType

    var shouldMerge: Boolean
}